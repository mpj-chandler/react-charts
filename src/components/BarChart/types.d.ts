import Placement from '../../enums/Placement';

export interface SeriesDataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullSeriesDataPoint extends SeriesDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    points: SeriesDataPoint[];
}

export interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
}
