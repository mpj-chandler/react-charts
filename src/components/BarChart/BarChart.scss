.BarChart {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-areas: "title title" "y-axis plot" "blank x-axis";
  grid-template-columns: 20% 60% 20%;
  grid-template-rows: 10% 70% 20%;

  &__title {
    grid-area: title;
  }

  &__blank {
    grid-area: blank;
    z-index: 0;
  }
}
