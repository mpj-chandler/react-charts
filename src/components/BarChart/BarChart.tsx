import React from 'react';
import styles from './BarChart.scss';
import XAxis from '../XAxis/XAxis';
import YAxis from '../YAxis/YAxis';
import { SeriesData, Padding } from './types';

export interface BarChartProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
}

const BarChart: React.FC<BarChartProps> = (props: BarChartProps) => {
    return (
        <div className={styles.BarChart}>
            <div className={styles.BarChart__title}>{props.title}</div>
            <YAxis title={'YAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__blank}>()</div>
            <XAxis title={'XAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__padRight}>()</div>
        </div>
    )
}

export default BarChart;
