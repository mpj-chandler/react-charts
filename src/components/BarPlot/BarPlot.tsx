import React from 'react';
import styles from './BarPlot.scss';
import { AxisConfig, Padding, SeriesData, SeriesDataPoint } from '../BarChart/types';
import Bar from '../Bar/Bar';
import { getYAxisRange } from '../../utils/getYAxisRange';
import { getXAxisRange } from '../../utils/getXAxisRange';
import Placement from '../../enums/Placement';

export interface BarPlotProps {
    padding: Padding;
    data: SeriesData[],
    xAxisConfig?: AxisConfig,
    yAxisConfig?: AxisConfig
}

const BarPlot: React.FC<BarPlotProps> = (props: BarPlotProps) => {
    const xRange = getXAxisRange(props.data, props.xAxisConfig);
    const yRange = getYAxisRange(props.data, props.yAxisConfig);
    const placement = props.xAxisConfig ? props.xAxisConfig.tickPlacement : Placement.Aligned;
    console.log('XRange: ', xRange, 'YRange: ', yRange);


    return (
        <svg className={styles.BarPlot}>
            <g>
                {props.data[0].points.map((point: SeriesDataPoint) => {
                    return (
                        <Bar
                            key={`${props.data[0].seriesName}${point.x}`}
                            point={point}
                            xRange={xRange}
                            yRange={yRange}
                            width={10}
                            fill={'red'}
                            stroke={'black'}
                            padding={props.padding}
                            numBars={props.data[0].points.length}
                            barAlignment={placement}
                        />
                    );
                })}
            </g>
        </svg>
    )
}

export default BarPlot;
