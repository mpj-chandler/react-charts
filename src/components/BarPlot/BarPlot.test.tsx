import React from 'react';
import TestRenderer from 'react-test-renderer';
import BarPlot, { BarPlotProps } from './BarPlot';
import { AxisConfig } from '../BarChart/types';
import Placement from '../../enums/Placement';

jest.mock('../Bar/Bar.tsx');

describe('BarPlot', () => {
    const inlineConfig: AxisConfig = {
        zeroIntercept: false,
        margin: 0,
        tickPlacement: Placement.Aligned
    };

    const bucketConfig: AxisConfig = { ...inlineConfig, tickPlacement: Placement.Bucket };

    const props: BarPlotProps = {
        data: [
            {
                seriesName: 'A',
                points: [
                    {
                        x: 0,
                        y: 10
                    },
                    {
                        x: 1,
                        y: 20
                    },
                    {
                        x: 2,
                        y: 30
                    }
                ]
            }
        ],
        padding: {
            top: 10,
            bottom: 10,
            left: 10,
            right: 10
        }
    }

    it('it renders consistently with defaults', () => {
        const component = TestRenderer.create(<BarPlot {...props}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with tick alignment explicitly set to Aligned for both axes', () => {
        const component = TestRenderer.create(<BarPlot {...props} xAxisConfig={inlineConfig} yAxisConfig={inlineConfig}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with tick alignment explicitly set to Aligned for XAxis and Bucket for YAxis', () => {
        const component = TestRenderer.create(<BarPlot {...props} xAxisConfig={inlineConfig} yAxisConfig={bucketConfig}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with tick alignment explicitly set to Bucket for XAxis and Aligned for YAxis', () => {
        const component = TestRenderer.create(<BarPlot {...props} xAxisConfig={bucketConfig} yAxisConfig={inlineConfig}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with tick alignment explicitly set to Bucket for both axes', () => {
        const component = TestRenderer.create(<BarPlot {...props} xAxisConfig={bucketConfig} yAxisConfig={bucketConfig}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });
})
