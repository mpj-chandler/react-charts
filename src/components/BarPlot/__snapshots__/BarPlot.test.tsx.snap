// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarPlot When all data points are non-null it renders consistently with defaults 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={0}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 0,
          "y": 10,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={1}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 1,
          "y": 20,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When all data points are non-null it renders consistently with tick alignment explicitly set to Aligned for XAxis and Bucket for YAxis 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={0}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 0,
          "y": 10,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={1}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 1,
          "y": 20,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When all data points are non-null it renders consistently with tick alignment explicitly set to Aligned for both axes 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={0}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 0,
          "y": 10,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={1}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 1,
          "y": 20,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When all data points are non-null it renders consistently with tick alignment explicitly set to Bucket for XAxis and Aligned for YAxis 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={1}
      fill="red"
      index={0}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 0,
          "y": 10,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={1}
      fill="red"
      index={1}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 1,
          "y": 20,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={1}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When all data points are non-null it renders consistently with tick alignment explicitly set to Bucket for both axes 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={1}
      fill="red"
      index={0}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 0,
          "y": 10,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={1}
      fill="red"
      index={1}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 1,
          "y": 20,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
    <div
      barAlignment={1}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.6666666666666665,
          "min": 0,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When some data points are null it renders consistently with defaults 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 1,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When some data points are null it renders consistently with tick alignment explicitly set to Aligned for XAxis and Bucket for YAxis 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 1,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When some data points are null it renders consistently with tick alignment explicitly set to Aligned for both axes 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={0}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2,
          "min": 1,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When some data points are null it renders consistently with tick alignment explicitly set to Bucket for XAxis and Aligned for YAxis 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={1}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.3333333333333335,
          "min": 1,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;

exports[`BarPlot When some data points are null it renders consistently with tick alignment explicitly set to Bucket for both axes 1`] = `
<svg
  className="BarPlot"
>
  <g>
    <div
      barAlignment={1}
      fill="red"
      index={2}
      numBars={3}
      padding={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      point={
        Object {
          "x": 2,
          "y": 30,
        }
      }
      stroke="black"
      type="NON_NULL_NUMERIC"
      width={10}
      xRange={
        Object {
          "max": 2.3333333333333335,
          "min": 1,
        }
      }
      yRange={
        Object {
          "max": 30,
          "min": 10,
        }
      }
    >
      Mock Bar 
    </div>
  </g>
</svg>
`;
